src/Tmds.Ssh/GssapiWithMicCredential.cs

```diff
@@ -0,0 +1,282 @@
+// This file is part of Tmds.Ssh which is released under MIT.
+// See file LICENSE for full license details.
+
+using System;
+using System.Buffers;
+using System.Linq;
+using System.Net;
+using System.Net.Security;
+using System.Reflection;
+using System.Security.Principal;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Tmds.Ssh;
+
+public sealed class GssapiWithMicCredential : Credential
+{
+    // gssapi-with-mic is defined in https://datatracker.ietf.org/doc/html/rfc4462
+    private const string AUTH_METHOD = "gssapi-with-mic";
+
+    // Kerberos - 1.2.840.113554.1.2.2
+    private static readonly byte[] KRB5_OID = [ 0x06, 0x09, 0x2A, 0x86, 0x48, 0x86, 0xF7, 0x12, 0x01, 0x02, 0x02 ];
+
+    private readonly Func<string?> _getPassword;
+    private readonly bool _delegateCredential;
+    private readonly string? _serviceName;
+
+#if NET8_0
+    private delegate void GetMICDelegate(ReadOnlySpan<byte> data, IBufferWriter<byte> writer);
+
+    private static readonly Lazy<MethodInfo?> _getMicMethInfo = new Lazy<MethodInfo?>(() =>
+        typeof(NegotiateAuthentication).GetMethod("GetMIC", BindingFlags.NonPublic | BindingFlags.Instance));
+
+    private MethodInfo GetMICMethod { get; }
+#endif
+
+    /// <summary>
+    /// Create a GSSAPI user authentication context with password string.
+    /// </summary>
+    /// <param name="password">The password to use for the GSSAPI username. Set to null to use a cached credential.</param>
+    /// <param name="delegateCredential">Request delegation on the GSSAPI context.</param>
+    /// <param name="serviceName">Override the service principal name (SPN), default it to use the connection hostname</param>
+    public GssapiWithMicCredential(string? password = null, bool delegateCredential = false, string? serviceName = null) : this(() => password, delegateCredential, serviceName)
+    { }
+
+    /// <summary>
+    /// Create a GSSAPI user authentication context with a password prompt.
+    /// </summary>
+    /// <param name="passwordPrompt">A prompting function that is called to retrieve the password when needed.</param>
+    /// <param name="delegateCredential">Request delegation on the GSSAPI context.</param>
+    /// <param name="serviceName">Override the service principal name (SPN), default it to use the connection hostname</param>
+    public GssapiWithMicCredential(Func<string?> passwordPrompt, bool delegateCredential = false, string? serviceName = null)
```

Comment:
> when needed

Looking at the implementation, it seems we're always prompting and not only when needed?

I think we should leave out the `passwordPrompt` and `password` arguments if that is the case.

Reply:
I mostly just followed the `PasswordCredential` setup (which also always prompts for a password). I assumed it was because a previous credential might be called before this one so the password may not have been prompted. Theoretically it can also fail if the server doesn't allow the auth OID that we are implementing (Kerberos).

It'll certainly be simpler to not have that and just accept a nullable string if that's what you would like here.

Reply:
> I think we should leave out the passwordPrompt and password arguments if that is the case.

Actually, we can keep the `password` arg.

Would it make sense to accept it as: `NetworkCredential? credential = null`?

Reply:
> I mostly just followed the PasswordCredential setup (which also always prompts for a password). I assumed it was because a previous credential might be called before this one so the password may not have been prompted. Theoretically it can also fail if the server doesn't allow the auth OID that we are implementing (Kerberos).

I would expect it to prompt only when needed. If it is always prompting, the user may provide the credential up-front.

For the password prompt, we actually skip the auth when the return value is `null`, while this is behaving different.

Let's leave out the `passwordPrompt` overload (for now).

Reply:
> Let's leave out the passwordPrompt overload (for now).

Makes sense!

> Would it make sense to accept it as: NetworkCredential? credential = null?

While it certainly could be used it does have a weird setup where there's the connection username specified and the one in the `NetworkCredential`. Which one would be used if we accepted a `NetworkCredential`, what would happen if they differ?

Reply:
I think we should use the one from the `NetworkCredential` and ignore the connection username.

Reply:
I'm looking at the .NET runtime implementation here:

https://github.com/dotnet/runtime/blob/f52856334d4929839ef97b7c11376ac51ebc6b2e/src/libraries/System.Net.Security/src/System/Net/NegotiateAuthenticationPal.Unix.cs#L194-L196

It seems that it is only considering the `UserName` when the `Password` is also set.

So, I think this makes sense:

If no credential is specified, assume it to be `CredentialCache.DefaultNetworkCredentials` and don't send a username in the `gssapi-with-mic` message.

If a credential is specified, it must include a username and password.

wdyt?

Reply:
> and don't send a username in the gssapi-with-mic message.

This depends on: _ The user name may be an empty string if it can be deduced from the
   results of the GSS-API authentication._. I'm not sure if this is the case?

Reply:
> I'm not sure if this is the case?

I gave it a try (against the test you've included), and setting an empty username doesn't work.

We can do the following: if the NetworkCredential.UserName is an empty string, use the connection username, and otherwise use the NetworkCredential.UserName.

Reply:
I just realized this username includes a domain too.

We can just start with an API that requires a `NetworkCredential`, and require a `UserName` to be set on this credential. The API can be relaxed later.

Reply:
> It seems that it is only considering the UserName when the Password is also set.

That is a good point. It is worth pointing out the path on Windows is a bit more convoluted but it calls [SspiEncodeStringsAsAuthIdentity](https://learn.microsoft.com/en-us/windows/win32/api/sspi/nf-sspi-sspiencodestringsasauthidentity) which doesn't actually fail on an empty password but I think that's a pretty rare scenario to think about on Windows.

> I gave it a try (against the test you've included), and setting an empty username doesn't work.

While the RFC states that the username can be null I found that openssh-portable explicitly fails if the username is not set https://github.com/openssh/openssh-portable/blob/a8fbe2f7d0d96d299ee8e69769e3b51067978748/auth2-gss.c#L114-L118. This happens after the first message which includes the username and OID the client supports and before the authentication token is first generated.

The same also applies to the Win32-OpenSSH fork which has the same code. This effectively means that we cannot rely on the cached credential, you must supply a user in some shape or form.

> I just realized this username includes a domain too.

At least on Linux, the Kerberos libraries have the concept of a default realm if configured in the `krb5.conf`. This means that `Environment.UserName` could still be possible in some scenarios.

Ultimately I think what we could do is accept `NetworkCredential?` which if `null` will default to `CredentialCache.DefaultNetworkCredential` for the GSSAPI auth stage. The `username` string supplied in the SSH messages should use `credential.UserName` with a fallback to the `settings.UserName` so an explicit value is always provided. This has the added benefit of being able to run as a specific target user (which may be different from our kerb principal) but authenticate with either the default kerb ticket or the explicit kerb credentials provided. It also means a realm may not be needed on the SSH connection string username.